src/content/en/fundamentals/engage-and-retain/push-notifications/common-notification-patterns.md

```diff
@@ -0,0 +1,375 @@
+[clients.openWindow()](https://developer.mozilla.org/en-US/docs/Web/API/Clients/openWindow)
+API.
+
+In our `notificationclick` event we'd run some kind like this:
+
+      const examplePage = '/demos/notification-examples/example-page.html';
+      const promiseChain = clients.openWindow(examplePage);
+      event.waitUntil(promiseChain);
+
+In the next section we'll look at how to check if the page we want to direct the user to is
+already open or not. This way we can focus on the open tab rather than constantly opening new
+tabs.
+
+## Focus an Existing Window
+
+When it's possible, we should focus a window rather than open a new window every time the user
+clicks a notification.
+
+Before we look at how to achieve this, it's worth highlighting that this is **only possible for
+pages on your origin**. This is because we can only see what pages are open that belong to our
+site, this prevents developers from being able to see all the sites their users are viewing.
+
+Taking the previous example where we opened
+'/demos/notification-examples/example-page.html', we'll alter it to see if it's already open or
+not.
+
+      const urlToOpen = new URL(examplePage, self.location.origin).href;
+
+      const promiseChain = clients.matchAll({
+        type: 'window',
+        includeUncontrolled: true
+      })
+      .then((windowClients) => {
+        let matchingClient = null;
+
+        for (let i = 0; i < windowClients.length; i++) {
+          const windowClient = windowClients[i];
+          if (windowClient.url === urlToOpen) {
+            matchingClient = windowClient;
+            break;
+          }
+        }
+
+        if (matchingClient) {
+          return matchingClient.focus();
+        } else {
+          return clients.openWindow(urlToOpen);
+        }
+      });
+
+      event.waitUntil(promiseChain);
+
+Let's step through the code.
+
+First we parse our example page using the URL API. This is a neat trick I picked up from [Jeff
+Posnick](https://twitter.com/jeffposnick). Calling `new URL()` with the `location` object will
+return an absolute URL if the string passed in is relative (i.e. '/' will become 'http://<Site
+Origin>/').
+
+We make the URL absolute so we can match it against the window URL's later on.
+
+      const urlToOpen = new URL(examplePage, self.location.origin).href;
+
+Then we get a list of the "WindowClients", which are the list of currently open tabs / windows
+(remember these are tabs for your origin only).
+
+      const promiseChain = clients.matchAll({
+        type: 'window',
+        includeUncontrolled: true
+      })
+
+The options passed into `matchAll` just inform the browser that we only want
+to search for "window type" clients (i.e. just look for tabs and windows. Exclude web workers).
+`includeUncontrolled` is to search to tabs that aren't controlled by the service worker this
+code is running in. Generally, you'll always want this to be true when using the API for
```

Comment:
Suggest:

is to search tabs from your origin that are _not_ controlled by your service worker.